RealmSwift/Realm.swift

```diff
@@ -1005,3 +1005,33 @@ extension Realm {
 
 /// The type of a block to run for notification purposes when the data in a Realm is modified.
 public typealias NotificationBlock = (_ notification: Realm.Notification, _ realm: Realm) -> Void
+
+@available(macOS 12.0, tvOS 15.0, iOS 15.0, watchOS 8.0, *)
+extension Realm {
+    private static let asyncOpenQueue = DispatchQueue(label: "io.realm.swift.asyncOpenDispatchQueue")
+    /**
+     Asynchronously open a Realm and deliver it to a block on the given queue.
+
+     Opening a Realm asynchronously will perform all work needed to get the Realm to
+     a usable state (such as running potentially time-consuming migrations) on a
+     background thread before dispatching back to the continiuation. In addition,
+     synchronized Realms wait for all remote content available at the time the
+     operation began to be downloaded and available locally.
+
+     - parameter configuration: A configuration object to use when opening the Realm.
+     - returns: An open Realm.
+     */
+    @discardableResult
+    public static func asyncOpen(configuration: Realm.Configuration = .defaultConfiguration) async throws -> Realm {
```

Comment:
Should this maybe just be `open()`? The fact that the function is `async` indicates that it's async, and I think the Swift team is hoping to avoid having function names with async prefixes.

Reply:
In that case, should we just scrap the static method and add an async init instead? Thoughts being that having a method named open (even if marked async) seems a bit odd to have alongside your standard init.

Suppose it would need to be well documented as a replacement for asyncOpen.

Reply:
Would that even be possible? I don't think you can overload on just asyncness.

Reply:
You can: https://github.com/apple/swift-evolution/blob/main/proposals/0296-async-await.md#overloading-and-overload-resolution

Swift resolves which version of the method to use based on if you are calling from an async context. So anybody opening a Realm in an async context would incidentally be using async open if we overloaded `init`.

Reply:
That seems like probably an issue because asyncOpen isn't just an async version of init. Maybe we should use this opportunity to rethink this a bit and make it so that it is? We could have `init(configuration: Realm.Config, behavior: DownloadBehavior = .openImmediately) async throws` so that `try! Realm()` and `await try! Realm()` do the same thing other than the async part, and then you'd pass `behavior: .alwaysDownloadLatest` or `behavior: .downloadOnFirstOpen` to opt in to other behavior.

(I think "download behavior" is what other SDKs are calling this but it's a really awkward name)